src/goto-programs/remove_exceptions.cpp

```diff
@@ -0,0 +1,598 @@
+                        null_pointer_exprt(pointer_typet()));
+    // jump to the next instruction
+    goto_programt::targett t_null=goto_program.insert_after(instr_it);
+    t_null->make_goto(next_it);
+    t_null->source_location=instr_it->source_location;
+    t_null->function=instr_it->function;
+    t_null->guard=eq_null;
+  }
+}
+
+/*******************************************************************\
+
+Function: remove_exceptionst::add_gotos
+
+Inputs:
+
+Outputs:
+
+Purpose: instruments each throw and function calls that may escape exceptions
+         with conditional GOTOS to the corresponding exception handlers
+
+\*******************************************************************/
+
+void remove_exceptionst::add_gotos(
+  const goto_functionst::function_mapt::iterator &func_it)
+{
+  // Stack of try-catch blocks
+  stack_catcht stack_catch;
+
+  goto_programt &goto_program=func_it->second.body;
+
+  if(goto_program.empty())
+    return;
+  Forall_goto_program_instructions(instr_it, goto_program)
+  {
+    // it's a CATCH but not a handler
+    if(instr_it->type==CATCH && !instr_it->code.has_operands())
+    {
+      if(instr_it->targets.empty()) // pop
+      {
+        // pop from the stack if possible
+        if(!stack_catch.empty())
+        {
+          stack_catch.pop_back();
+        }
+        else
+        {
+#ifdef DEBUG
+          std::cout << "Remove exceptions: empty stack" << std::endl;
+#endif
+        }
+      }
+      else // push
+      {
+        remove_exceptionst::catch_handlerst exception;
+        stack_catch.push_back(exception);
+        remove_exceptionst::catch_handlerst &last_exception=
+          stack_catch.back();
+
+        // copy targets
+        const irept::subt &exception_list=
+          instr_it->code.find(ID_exception_list).get_sub();
+        assert(exception_list.size()==instr_it->targets.size());
+
+        // Fill the map with the catch type and the target
+        unsigned i=0;
+        for(auto target : instr_it->targets)
+        {
+          last_exception.push_back(
+            std::make_pair(exception_list[i].id(), target));
+          i++;
+        }
+      }
+
+      instr_it->make_skip();
+    }
+    else if(instr_it->type==THROW)
+    {
+      add_throw_gotos(func_it, instr_it, stack_catch);
+    }
+    else if(instr_it->type==FUNCTION_CALL)
+    {
+      add_function_call_gotos(func_it, instr_it, stack_catch);
+    }
+  }
+}
+
+
+/*******************************************************************\
+
+Function: remove_exceptionst::operator()
+
+Inputs:
+
+Outputs:
+
+Purpose:
+
+\*******************************************************************/
+
+void remove_exceptionst::operator()(goto_functionst &goto_functions)
+{
+  Forall_goto_functions(it, goto_functions)
+  {
+    add_exceptional_returns(it);
+  }
+  Forall_goto_functions(it, goto_functions)
+  {
+    instrument_exception_handlers(it);
+    add_gotos(it);
+    instrument_function_calls(it);
+    replace_throws(it);
```

Comment:
I don't think I fully understand the prescribed sequence of actions here. It seems that `add_gotos` handles multiple kinds of instructions, whereas the others each iterate over the full goto program to just find and process a single kind of instructions. Is that necessary?